tasks/Task.cpp

```diff
@@ -2,48 +2,331 @@
+
+    peer_connection->onLocalDescription(
+        [&, protocol, remote_peer_id](rtc::Description description)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
+            message["protocol"] = protocol;
+            message["to"] = remote_peer_id;
+            message["actiontype"] = description.typeString();
+            message["description"] = string(description);
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+                ws->send(fast.write(message));
+        });
+
+    peer_connection->onLocalCandidate(
+        [&, protocol, remote_peer_id](rtc::Candidate candidate)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
+            message["protocol"] = protocol;
+            message["to"] = remote_peer_id;
+            message["actiontype"] = "candidate";
+            message["candidate"] = string(candidate);
+            message["mid"] = candidate.mid();
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+                ws->send(fast.write(message));
+        });
+
+    return peer_connection;
+}
+
+void Task::configurePeerDataChannel(string const &remote_peer_id)
 {
+    mPeerConnection->onDataChannel(
+        [&](shared_ptr<rtc::DataChannel> data_channel)
+        {
+            LOG_INFO_S << "DataChannel from " << remote_peer_id
+                       << " received with label \"" << data_channel->label() << "\""
+                       << std::endl;
+
+            mDataChannel = data_channel;
+            data_channel->onOpen([&, wdc = make_weak_ptr(data_channel)]() {});
+
+            data_channel->onMessage(
+                [&](variant<binary, string> data)
+                {
+                    RawPacket dataPacket;
+                    dataPacket.time = base::Time::now();
+
+                    if (holds_alternative<string>(data))
+                    {
+                        string data_string = get<string>(data);
+                        vector<uint8_t> new_data(data_string.begin(), data_string.end());
+                        dataPacket.data.resize(new_data.size());
+                        dataPacket.data = new_data;
+                    }
+                    else
+                    {
+                        vector<byte> data_byte = get<binary>(data);
+                        dataPacket.data.resize(data_byte.size());
+                        for (unsigned int i = 0; i < data_byte.size(); i++)
+                        {
+                            dataPacket.data[i] = to_integer<uint8_t>(data_byte[i]);
+                        }
+                    }
+                    _data_out.write(dataPacket);
+                });
+        });
 }
 
-Task::~Task()
+void Task::configureWebSocket()
 {
+    promise<void> ws_promise;
+    auto ws_future = ws_promise.get_future();
+
+    mWs->onOpen(
+        [&ws_promise]()
+        {
+            LOG_INFO_S << "WebSocket connected, signaling ready" << std::endl;
+            ws_promise.set_value();
```

Comment:
Same ... should have some status field set when the websocket is opened and when it closes.